sklearn/manifold/t_sne.py

```diff
@@ -433,34 +690,73 @@ def fit(self, X, y=None):
         else:
             if self.verbose:
                 print("[t-SNE] Computing pairwise distances...")
-
             if self.metric == "euclidean":
-                distances = pairwise_distances(X, metric=self.metric, squared=True)
+                distances = pairwise_distances(X, metric=self.metric,
+                                               squared=True)
             else:
                 distances = pairwise_distances(X, metric=self.metric)
+        distances = check_array(distances, dtype=np.float32)
+
+        if not np.all(distances >= 0):
+            raise ValueError("All distances should be positive, either "
+                             "the metric or precomputed distances given "
+                             "as X are not correct")
 
         # Degrees of freedom of the Student's t-distribution. The suggestion
-        # alpha = n_components - 1 comes from "Learning a Parametric Embedding
-        # by Preserving Local Structure" Laurens van der Maaten, 2009.
-        alpha = max(self.n_components - 1.0, 1)
+        # degrees_of_freedom = n_components - 1 comes from
+        # "Learning a Parametric Embedding by Preserving Local Structure"
+        # Laurens van der Maaten, 2009.
+        degrees_of_freedom = max(self.n_components - 1.0, 1.0)
         n_samples = X.shape[0]
         self.training_data_ = X
+        # the number of nearest neighbors to find
+        k = min(n_samples - 1, int(3. * self.perplexity + 1))
+
+        neighbors_nn = None
+        if self.method == 'barnes_hut':
+            if self.verbose:
+                print("[t-SNE] Computing %i nearest neighbors..." % k)
+            if self.metric == 'precomputed':
+                # Use the precomputed distances to find
+                # the k nearest neighbors and their distances
+                neighbors_nn = np.argsort(distances, axis=1)[:, :k]
+            else:
+                # Find the nearest neighbors for every point
+                bt = BallTree(X)
+                # LvdM uses 3 * perplexity as the number of neighbors
+                # And we add one to not count the data point itself
+                # In the event that we have very small # of points
+                # set the neighbors to n - 1
+                distances_nn, neighbors_nn = bt.query(X, k=k+1)
+                neighbors_nn = neighbors_nn[:, 1:]
```

Comment:
This won't be correct if a data point is duplicated. A correct version of this functionality is now provided by `NearestNeighbors(...).fit(X).kneighbors()` (don't pass `X` to `kneighbors` and it will query with the indexed data, excluding the query point from the results).